doc/tutorial/hello_world.dox

```diff
@@ -0,0 +1,165 @@
+ * we're going to be creating publisher and subscriber nodes.
+ *
+ * Creating a node requires three parameters: the topic separators, a
+ * key store, and a key identifier.  For now we're only concerned with
+ * the separators.  Key stores and identifiers are covered later when
+ * discussing how to secure communications.
+ *
+ * The separators parameter is a string containing the characters used
+ * as topic level separators.  A topic is composed of multiple levels
+ * with separators between them.  Providing @c /. as the separators
+ * parameter value means all of the following topics are equivalent:
+ * @code
+ * a/b/c/d
+ * a.b.c.d
+ * a/b.c/d
+ * @endcode
+ *
+ * @see DPS_SetNodeData(), DPS_GetNodeData()
+ *
+ * @section starting-a-node Starting a node
+ * @snippet tutorial.c Starting a node
+ *
+ * Once created, a node must be started.  Starting a node enables it
+ * to begin sending and receiving DPS messages in the network.
+ *
+ * For this example, we are going to be sending and receiving
+ * multicast publications so we enable both and let DPS assign the
+ * listening port.
+ *
+ * @see DPS_MCAST_PUB_DISABLED, DPS_GetPortNumber()
+ *
+ * @section publishing Publishing
+ *
+ * @subsection creating-a-publication Creating a publication
+ * @snippet tutorial.c Creating a publication
+ *
+ * Each publication in DPS is represented by a @c DPS_Publication.
+ * Each publication has a set of topics, a UUID, and a sequence
+ * number.  In this example we are creating a publication with one
+ * topic, @c a/b/c/d.  The UUID is assigned by DPS and the sequence
+ * number will be incremented each time we publish.
+ *
+ * The @c noWildCard parameter is used by the publisher to control
+ * whether a subscription is required to match the publication's
+ * topics exactly or can use wildcards to match the topics.  If we set
+ * @c noWildCard to @c DPS_TRUE then only a subscription to @c a/b/c/d
+ * will receive this publication.  Since we set @c noWildCard to @c
+ * DPS_FALSE here, subscriptions to @c a/#, @c a/+/+/d, or similar
+ * variations will receive this publication.
+ *
+ * Both the publication's key identifier and acknowledgement handler
+ * are set to @c NULL here; they are covered in later sections.
+ *
+ * @see DPS_SetPublicationData(), DPS_GetPublicationData(),
+ *      DPS_PublicationGetNode(), DPS_PublicationGetUUID(),
+ *      DPS_PublicationGetSequenceNum()
+ *
+ * @subsection sending-a-publication Sending a publication
+ * @snippet tutorial.c Sending a publication
+ *
+ * Once created and initialized with a set of topics, application
+ * payloads may be sent.  Payload data is simply an array of bytes in
+ * DPS, no assumptions are made with regards to the payload format.
+ *
+ * In this example the @c ttl parameter is zero, indicating that the
+ * publication will be sent best-effort to all active subscribing
+ * nodes.  A non-zero ttl is referred to as a retained publication
+ * and is covered later.
```

Comment:
I the publication fails to send to a node is it completely lost? Is there any retry?

The way I read this the message is published to all subscribers using a single attempt (best-effort) and is then discarded. The publisher does not actually know if the subscribers got the publication right?

Reply:
Your understanding is correct.  There is no retry.  The publisher doesn't necessarily need to know who the subscribers are.  If it does, then it needs to use acknowledgements.

Reply:
Unlike client/server systems where the client knows the server and visa versa in pub/sub systems the publishers don't know if there are any subscribers and the subscribers only know about publishers that publish topics that match the subscriber's. This so-called loose coupling is one of the big advantages of pub/sub. As Todd mentions, DPS provides an acknowledgement mechanism that allows a publisher to request that subscribers send a reply message. Subscribers are not required to send a reply but a publisher can choose to resend a publication if it requests a reply and doesn't get one. This ACK mechanism can be used to implement RPC or command/response behavior or when multiple replies are required or expected can support more exotic uses cases such as weighted decision making, voting, and high availability.